core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,226 @@
+package io.grpc;
+
+import com.google.common.base.Preconditions;
+
+import java.io.Closeable;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+
+import javax.annotation.Nullable;
+
+
+/**
+ * --Add warning about using this too liberally--
+ * --Add warning about using this with containers that unload Classloaders--
+ *
+ *
+ */
+// Discussion point. Using java.io.Closeable here so we can inherit AutoCloseable where its available.
+public final class Context implements Closeable {
+
+  /**
+   * The logical root context which is {@link #current()} if no other context is bound. This context
+   * is already closed and will not retain any listeners.
+   */
+  public static final Context ROOT = new Context();
+  static {
+    ROOT.close();
```

Comment:
Not sure I quite follow, as currently defined withListener() only adds a listener to the parent and does not propagate any further up. If I call withListener and the parent is already closed then the listener is notified immediately (closure cascades by default and is immediate in this case) and the parent is not mutated.